src/H5Dmpio.c

```diff
@@ -2927,187 +3006,258 @@ H5D__obtain_mpio_mode(H5D_io_info_t *io_info, H5D_dset_io_info_t *di, uint8_t as
+                 * twice, using just the relevant information from the message before and after I/O.
+                 * The better way would be to avoid processing chunk messages twice by extracting (and
+                 * keeping around) the dataspace portion of the message before I/O and processing the
+                 * rest of the chunk message after I/O. Note that the dataspace portion of each chunk
+                 * message is used to correctly apply chunk modification data from the message, so
+                 * must be kept around both before and after I/O in this case.
+                 */
+                if (io_info->op_type == H5D_IO_OP_READ)
+                    local_info_array[buf_idx].need_read = true;
+                else {
+                    local_info_array[buf_idx].need_read =
+                        local_info_array[buf_idx].io_size <
+                        (size_t)di[dset_idx].dset->shared->layout.u.chunk.size;
+                }
+
+                if (local_info_array[buf_idx].need_read)
+                    num_chunks_to_read++;
+
+                local_info_array[buf_idx].skip_filter_pline = false;
+                if (!filter_partial_edge_chunks) {
+                    /*
+                     * If this is a partial edge chunk and the "don't filter partial edge
+                     * chunks" flag is set, make sure not to apply filters to the chunk.
+                     */
+                    if (H5D__chunk_is_partial_edge_chunk(
+                            di[dset_idx].dset->shared->ndims, di[dset_idx].dset->shared->layout.u.chunk.dim,
+                            chunk_info->scaled, di[dset_idx].dset->shared->curr_dims))
+                        local_info_array[buf_idx].skip_filter_pline = true;
+                }
+
+                /* Initialize the chunk's shared info */
+                local_info_array[buf_idx].chunk_current = udata.chunk_block;
+                local_info_array[buf_idx].chunk_new     = udata.chunk_block;
 
-            local_info_array[i].skip_filter_pline = false;
-            if (!filter_partial_edge_chunks) {
                 /*
-                 * If this is a partial edge chunk and the "don't filter partial edge
-                 * chunks" flag is set, make sure not to apply filters to the chunk.
+                 * Check if the list is not in ascending order of offset in the file
+                 * or has unallocated chunks. In either case, the list should get
+                 * sorted.
                  */
-                if (H5D__chunk_is_partial_edge_chunk(di->dset->shared->ndims,
-                                                     di->dset->shared->layout.u.chunk.dim, chunk_info->scaled,
-                                                     di->dset->shared->curr_dims))
-                    local_info_array[i].skip_filter_pline = true;
-            }
+                if (buf_idx) {
```

Comment:
Might be worth changing the condition to if (!need_sort && buf_idx). Can't be certain without looking more investigation though (not sure how heavyweight the block is when compiled)

Reply:
I'd guess it wouldn't be heavy enough to make much of a difference, but I think making that change is still a good idea.